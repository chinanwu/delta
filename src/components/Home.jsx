import React, { useCallback, useEffect, useState } from 'react';
import { Redirect } from 'react-router-dom';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { postFetch } from '../functions/FetchFunctions';
import generateGameUrl from '../functions/generateGameUrl';
import getThemeClassname from '../functions/getThemeClassname';
import hasValidCharacters from '../functions/hasValidCharacters';

import './Home.less';
import NavBar from './NavBar.jsx';

export const Home = ({ dark }) => {
	const [gameUrl, setGameUrl] = useState(generateGameUrl);
	const [error, setError] = useState(null);
	const [redirect, setRedirect] = useState(false);

	useEffect(() => {
		document.title = `Home - ${document.title}`;
	}, []);

	const handleChange = useCallback(
		event => {
			if (event && event.target) {
				if (event.target.value) {
					const url = event.target.value;
					setGameUrl(url);
					hasValidCharacters(url)
						? setError(null)
						: setError('Game URL must only contain letters');
				} else {
					setGameUrl('');
					setError('Game URL cannot be empty');
				}
			}
		},
		[setGameUrl, setError]
	);

	const handleCreateClick = useCallback(() => {
		!error
			? postFetch(
					'http://localhost:5000/api/games/new',
					JSON.stringify({ url: gameUrl })
			  ).then(res => {
					console.log('Creating game: ' + res.success);
					if (res.success) {
						setRedirect(true);
					}
			  })
			: null;
	}, [error, setRedirect]);

	return redirect ? (
		<Redirect to={'/game/' + gameUrl} />
	) : (
		<div className={getThemeClassname('Home', dark)}>
			<NavBar activeTab="home" />
			<div className="Home__header" role="banner">
				Mairead
			</div>
			<div>
				<div className="Home__game">
					<input
						id="homeInput"
						className={'Home__input' + (error ? ' Home__input--error' : '')}
						type="text"
						name="gameUrl"
						value={gameUrl}
						onChange={handleChange}
					/>
					<button
						className={'Home__btn' + (error ? ' Home__btn--error' : '')}
						disabled={!!error}
						onClick={handleCreateClick}
					>
						Create
					</button>
				</div>
				<div className="Home__error">{error}</div>
			</div>
			<div
				className={getThemeClassname('Home__footer', dark)}
				aria-label="Footer"
			>
				Made with love by{' '}
				<a href="https://www.github.com/chinanwu">Chin-An Wu</a>
			</div>
		</div>
	);
};

Home.propTypes = {
	dark: PropTypes.bool,
};

export const mapStateToProps = ({ theme }) => ({
	dark: theme.dark,
});

export default connect(mapStateToProps)(Home);
