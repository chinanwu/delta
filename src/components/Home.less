@import '../style/styles';

.Home {
	display: flex;
	flex-direction: column;
	justify-content: center;
	align-items: center;
	height: 100%;
}

.Home__nav {
	position: absolute;
	top: 24px;
	left: 12px;
	font-size: 18px;
}

.Home__about {
	padding: 12px;
	border-radius: 2px;
	background-color: @blackBtn;
	box-shadow: @box-shadow-01;
	color: white;
}

.Home__about:hover {
	background-color: @blackBtn-light;
	box-shadow: @box-shadow-02;
}

.Home__about:active {
	background-color: @blackBtn;
	box-shadow: @box-shadow-02;
}

.Home__header {
	font-size: 64px;
	margin-bottom: 32px;
}

.Home__game {
	display: flex;
	margin-bottom: 12px;
}

.Home__input {
	padding-top: 12px;
	padding-bottom: 12px;
	font-size: 26px;
	margin-right: 12px;
	text-align: center;
	box-shadow: @box-shadow-01;
	border: 1px solid black;
}

.Home__input--error {
	border: 2px solid @errorRed;
}

.home__btn {
	background-color: @greenBtn;
	border: none;
	padding-top: 12px;
	padding-bottom: 12px;
	font-size: 26px;
	box-shadow: @box-shadow-01;
}

.home__btn:hover {
	background-color: @greenBtnHover;
	box-shadow: @box-shadow-02;
}

.home__btn:active {
	background-color: @greenBtnActive;
}

.home__btn--error {
	background-color: @disabledBtn;
	cursor: not-allowed;
}

.home__btn--error:hover {
	background-color: @disabledBtn;
}

.home__btn--error:active {
	background-color: @disabledBtn;
}

.Home__error {
	position: absolute;
}

.Home__footer {
	position: absolute;
	bottom: 0;
}
