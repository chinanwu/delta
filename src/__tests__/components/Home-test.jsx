import { shallow } from 'enzyme';
import React from 'react';

import { Home, mapStateToProps } from '../../components/Home.jsx';

jest.unmock('../../components/Home.jsx');

describe('Home component', () => {
	describe('rendering', () => {
		it('renders Home', () => {
			const wrapper = shallow(<Home />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('mapStateToProps', () => {
		it('returned mapped properties', () => {
			const theme = { dark: true };
			const state = { theme };
			expect(mapStateToProps(state)).toEqual(theme);
		});
	});
});
