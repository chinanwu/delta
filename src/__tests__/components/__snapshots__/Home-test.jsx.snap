// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component rendering renders Home 1`] = `
<div>
  <Memo(Connect(NavBar))
    activeTab="home"
  />
  <div
    className="Home__header"
    role="banner"
  >
    Mairead
  </div>
  <div>
    <div
      className="Home__game"
    >
      <input
        className="Home__input"
        id="homeInput"
        name="gameUrl"
        onChange={[Function]}
        type="text"
        value={[MockFunction]}
      />
      <button
        className="Home__btn"
        disabled={false}
        onClick={[Function]}
      >
        Create
      </button>
    </div>
    <div
      className="Home__error"
    />
  </div>
  <div
    aria-label="Footer"
  >
    Made with love by
     
    <a
      href="https://www.github.com/chinanwu"
    >
      Chin-An Wu
    </a>
  </div>
</div>
`;
